### PR TITLE
Import BatchMatMul to ONNX

### PR DESCRIPTION
**Pull request by yinghai**
*Description*:
Import  BatchMatMul to ONNX
*Testing*:
unittest
*Documentation*:
Fixes https://github.com/pytorch/glow/issues/1970

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

### CODE REVIEW

File: include/glow/Importer/CommonOperatorLoader.h [hunk 1]
```diff
@@ -188,6 +188,50 @@ class CommonOperatorLoader : public ProtobufLoader {
...
   }
 
+  void loadBatchMatMul(const OpType &op, ArgumentDictionaryTy &dict) {
+    const std::string &opName = loadOperatorName(op);
+    auto LHS = getNodeValueOrCreateConstantByName(op.input(0));
+    auto RHS = getNodeValueOrCreateConstantByName(op.input(1));
+
+    bool transLHS = dict.count("trans_a") && (loadInt(dict["trans_a"]) == 1);
+    (void)transLHS;
+    assert(!transLHS && "Don't support transpose lhs for now.");
```
Comment (rdzhabarov) [inline include/glow/Importer/CommonOperatorLoader.h]: can you reflect this properly, e.g., returning false

File: include/glow/Importer/CommonOperatorLoader.h [hunk 2]
```diff
@@ -188,6 +188,50 @@ class CommonOperatorLoader : public ProtobufLoader {
...
+    bool transRHS = dict.count("trans_b") && (loadInt(dict["trans_b"]) == 1);
+    if (transRHS) {
+      // The semantic of the transpose in that context is:
+      // swap the last two dimensions.
+      unsigned_t nbDims = RHS.dims().size();
+      GLOW_ASSERT(nbDims >= 2 && "C2 specs say rank of RHS must be >= 2");
```
Comment (rdzhabarov) [inline include/glow/Importer/CommonOperatorLoader.h]: also return false

File: include/glow/Importer/CommonOperatorLoader.h [hunk 3]
```diff
@@ -556,6 +600,10 @@ class CommonOperatorLoader : public ProtobufLoader {
       loadReduceMeanOrSum(typeName, op, dict);
       return true;
     }
+    if (typeName == "BatchMatMul") {
+      loadBatchMatMul(op, dict);
```
Comment (rdzhabarov) [inline include/glow/Importer/CommonOperatorLoader.h]: return loadBatchMatMul() after that methods is refactored.

File: tests/unittests/onnxImporterTest.cpp [hunk 4]
```diff
@@ -133,6 +133,68 @@ TEST(onnx, importClip) {
...
+  auto &mod = EE.getModule();
+  Function *F = mod.createFunction("main");
+  std::string netFilename("tests/models/onnxModels/batch_matmul.onnxtxt");
+
+  Context ctx;
+  Placeholder *output;
+  {
+    Tensor inputs_0(ElemKind::FloatTy, {20, 7, 40});
+    Tensor inputs_1(ElemKind::FloatTy, {20, 7, 40});
+    inputs_0.getHandle().randomize(-3.0, 3.0, mod.getPRNG());
```
Comment (rdzhabarov) [inline tests/unittests/onnxImporterTest.cpp]: no need to init inputs.

File: tests/unittests/onnxImporterTest.cpp [hunk 5]
```diff
@@ -133,6 +133,68 @@ TEST(onnx, importClip) {
...
+    inputs_1.getHandle().randomize(-3.0, 3.0, mod.getPRNG());
+    ONNXModelLoader onnxLD(netFilename, {"inputs_0", "inputs_1"},
+                           {&inputs_0.getType(), &inputs_1.getType()}, *F);
+    output = onnxLD.getSingleOutput();
+
+    ctx.allocate(mod.getPlaceholders());
+    updateInputPlaceholdersByName(ctx, &mod, {"inputs_0", "inputs_1"},
```
Comment (rdzhabarov) [inline tests/unittests/onnxImporterTest.cpp]: no need to update inputs, we do not check results anyway.

File: tests/unittests/onnxImporterTest.cpp [hunk 6]
```diff
@@ -133,6 +133,68 @@ TEST(onnx, importClip) {
...
+  auto *res = ctx.get(output);
+  EE.compile(CompilationMode::Infer, F, ctx);
+  EE.run(ctx);
+
+  auto result = res->getHandle();
+  std::vector<size_t> expectedDims = {20, 7, 7};
+  EXPECT_EQ(result.dims().vec(), expectedDims);
+
+  // High level check on the content of the graph.
+  // We have 1 transpose, 20 * (matmul + 2 slices, 2 reshapes), 1 concat, 1
```
Comment (rdzhabarov) [inline tests/unittests/onnxImporterTest.cpp]: nit: matmul, 2 slices, 2 reshapes

File: include/glow/Importer/CommonOperatorLoader.h (hunk 1 above)
Comment (yinghai) [reply include/glow/Importer/CommonOperatorLoader.h]: I can return `false` but how am I going to clearly issue the error message?

**Comment by yinghai** [comment]
@rdzhabarov Looking through the code in `CommonOperatorLoader.h`, we have been consistently using `assert` plus error message to do checks. Not that I agree with this approach, I think for this PR we should stick to this convention and do it in one shot in some refactoring PR (https://github.com/pytorch/glow/issues/1517). cc: @qcolombet

**Comment by yinghai** [comment]
Yeah, we should follow this PR https://github.com/pytorch/glow/pull/1993

**Comment by rdzhabarov** [comment]
Verifier checks are too late, we need to report errors on multiple levels.

Verification on inputs are specifically important because we'd detect those before compiling the model. Feel free to defer checks for later.

File: include/glow/Importer/CommonOperatorLoader.h (hunk 1 above)
Comment (artemrakhov) [reply include/glow/Importer/CommonOperatorLoader.h]: This is a different story. I suggest not to worry about it in this PR.

File: lib/Importer/Caffe2ModelLoader.cpp [hunk 7]
```diff
@@ -476,7 +476,7 @@ void Caffe2ModelLoader::loadOperator(const caffe2::OperatorDef &op) {
     return;
   }
 
-  if (typeName == "MatMul" || typeName == "BatchMatMul") {
+  if (typeName == "MatMul") {
```
Comment (artemrakhov) [inline lib/Importer/Caffe2ModelLoader.cpp]: You can unify loading of MatMul and BatchMatMul by introducing a flag `isBatched` to `loadBatchMatMul()` method, and calling `loadBatchMatMul(..., isBatched=false)` method from here. In other words, regular `MatMul` won't fall under common ops, but will still share code.

File: lib/Importer/ONNXIFIModelLoader.cpp [hunk 8]
```diff
@@ -197,6 +197,12 @@ ONNXIFIModelLoader::parseOperators(const void *onnxModel,
       ADD_OP_MAPPING(MinNodeKind, FloatTy);
       ADD_OP_MAPPING(MaxNodeKind, FloatTy);
       ADD_OP_MAPPING(SplatNodeKind, FloatTy);
+    } else if (operation == "BatchMatMul") {
+      ADD_OP_MAPPING(TransposeNodeKind, FloatTy);
```
Comment (artemrakhov) [inline lib/Importer/ONNXIFIModelLoader.cpp]: In some cases `BatchMatMul` is just one `MatMulNode`. In some cases there's just no transpose. How can we account for this?

**Comment by yinghai** [comment]
Oh, what I meant is to use similar check and error reporting mechanism instead of assert.

File: lib/Importer/ONNXIFIModelLoader.cpp (hunk 8 above)
Comment (yinghai) [reply lib/Importer/ONNXIFIModelLoader.cpp]: Currently the strategy is to cover the max-set which we support anyway. `BatchMatMul` is really a hack from ONNX point of view, maybe we can export it more appropriately, but this is not a priority for now.

Comment (rdzhabarov) [reply lib/Importer/ONNXIFIModelLoader.cpp]: Tracking this is 1950 issue

**Comment by artemrakhov** [comment]
LGTM

**Comment by rdzhabarov** [comment]
LG! Thanks

**Pull request merged**